design/src/aws_config.md

```diff
@@ -0,0 +1,161 @@
+# AWS Configuration RFC
+
+> Status: RFC. For an ordered list of proposed changes see: [Proposed changes](#changes-checklist).
+
+An AWS SDK loads configuration from multiple locations. Some of these locations can be loaded synchronously. Some are
+async. Others may actually use AWS services such as STS or SSO.
+
+This document proposes an overhaul to the Config design to facilitate three things:
+
+1. Future-proof: It should be easy to add additional sources of region and credentials, sync and async, from many
+   sources, including code generated services.
+2. Ergonomic: There should be one obvious way to create an AWS service client. Customers should be able to easily
+   customize the client to make common changes. It should lead customers into the pit of success in terms of sharing
+   things that are expensive to create.
+3. Shareable: A config object should be usable to configure multiple AWS services.
+
+## New customer experience
+
+With the new shared config, customers must always take `aws-config` as a dependency if they want access to the automatic
+configuration loading.
+
+In Cargo.toml:
+
+```toml
+[dependencies]
+aws-sdk-dynamo = "0.1"
+aws-config = "0.5"
+```
+
+When creating a client, customers will create a `SharedConfig` struct, then use it to construct the service client.
+The `SharedConfig` struct is defined centrally in `aws-types`. (See [stability](#stability-and-versioning))
+
+```rust
+// pub use of aws_types::SharedConfig
+use aws_config::SharedConfig;
+
+async fn main() {
+    // loading a shared config may make network calls, etc.
+    // with some custom configuration:
+    let config = aws_config::config_loader()
+            .region(Region::new("us-east-1"))
+            .sleep(async_std_sleep)
```

Comment:
I know this doesn't commit to an API, but I'd maybe use `every` instead of `sleep`, since I'm not sure what is supposed to be sleeping. Alternatively: this can be an stream that yields every `Duration` with new configuration?

Reply:
Yeah, `sleep` isn't the most clear name for this. This is for supporting multiple async runtimes. We need a way to wire up the correct runtime's async sleep operation so that if, for example, you wanted to use async-std rather than Tokio, you could just wire it up here (or more realistically, it would get wired up automatically by the Cargo features and this would only be used if you were doing something special).

We could change this to `async_runtime(..)` and make a larger trait that accounts for all of the async runtime differences, but that could be a backwards compatibility issue if we need to add new operations to it. Right now, we only need sleep (to support timeouts). It's unclear what else we'll need in the future.